// @aws-sdk/client-sagemaker/dist-es/models/models_5.js
import { SENSITIVE_STRING } from "@smithy/smithy-client";
import { ModelPackageModelCardFilterSensitiveLog, } from "./models_1";
import { OidcConfigFilterSensitiveLog, } from "./models_2";
import { ModelCardFilterSensitiveLog, ModelPackageFilterSensitiveLog, } from "./models_4";
export const SearchSortOrder = {
    ASCENDING: "Ascending",
    DESCENDING: "Descending",
};
export const VariantPropertyType = {
    DataCaptureConfig: "DataCaptureConfig",
    DesiredInstanceCount: "DesiredInstanceCount",
    DesiredWeight: "DesiredWeight",
};
export const SearchRecordFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.TrialComponent && { TrialComponent: obj.TrialComponent }),
    ...(obj.ModelPackage && { ModelPackage: ModelPackageFilterSensitiveLog(obj.ModelPackage) }),
    ...(obj.FeatureGroup && { FeatureGroup: obj.FeatureGroup }),
    ...(obj.ModelCard && { ModelCard: ModelCardFilterSensitiveLog(obj.ModelCard) }),
});
export const SearchResponseFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.Results && { Results: obj.Results.map((item) => SearchRecordFilterSensitiveLog(item)) }),
});
export const UpdateModelCardRequestFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.Content && { Content: SENSITIVE_STRING }),
});
export const UpdateModelPackageInputFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.ModelCard && { ModelCard: ModelPackageModelCardFilterSensitiveLog(obj.ModelCard) }),
});
export const UpdateWorkforceRequestFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.OidcConfig && { OidcConfig: OidcConfigFilterSensitiveLog(obj.OidcConfig) }),
});
